docs/developer_guide.md

```diff
@@ -1,8 +1,147 @@
+
+# ----- Your Robot specific tools -----
+@tool
+def eat_grass():
+    """Eat grass tool"""
+    return cow.eat_grass()
+
+@tool
+def spy_on_farmer():
+    """Spy on farmer tool"""
+    return cow.spy_on_farmer()
+
+@tool
+def return_to_base():
+    """Return to base tool"""
+    return cow.return_to_base()
+
+@tool
+def run_away():
+    """Run away tool"""
+    return cow.run_away()
+
+def state_retriever():
+    """State retriever used for feeding state information
+    to agent every iteration"""
+    return {"temperature": 30, "battery_state": 33}
+
+# ----- Initialize -----
+llm = ChatOpenAI(model="gpt-4o")
+tools = [eat_grass, spy_on_farmer, return_to_base, run_away]
+agent = create_state_based_agent(llm=llm, state_retriever=state_retriever, tools=tools)
+system_prompt = "You are a cow. You are in a field. "
+user_prompt = (
+    "Please eat the grass 3 times as a cover, then run away at first possibility."
+    "Make sure that you are not caught by the farmer."
+)
+
+# ----- Invoke the agent -----
+output = agent.invoke({"messages": [system_prompt, user_prompt]}, config={"recursion_limit": 50})
+
+print(output["messages"][-1].json(indent=4))
+```
+
+## Adjusting RAI for Your Robot
+
+### 1. Create Your Robot whoami configuration package [Optional]
+
+The whoami package is used for configuring your robot within the RAI ecosystem. It helps define the robot's identity, self-understanding, ethical code, and documentation. For instructions see [configure RAI for your robot](create_robots_whoami.md).
+
+> [!NOTE]  
+> While RAI's agents can be started without a whoami configuration package, we strongly recommend integrating it for a more complete setup.
+
+### 2. Implement new tools specific to your robot
+
+To extend RAI with tools specific to your robot, implement new tool functions as shown below. These functions should be decorated with @tool and should interact with your robot's API.
+
+```python
+from langchain.tools import tool, BaseTool
+from langchain.pydantic_v1 import BaseModel, Field
+from typing import Type
+from myrobot import robot
+
+# decorator based api, for basic tools
+@tool
+def pick_up_object(bbox_3d):
+    """Tool used for picking up objects"""
+    return robot.pick_up_object(bbox_3d)
+
+@tool
+def scan_object():
+    """Tool used for x-raying previously picked object"""
+    return robot.scan_object()
+
+# class based api, useful when tools use foreign objects like ROS 2 Node
+class SayToolInput(BaseModel):
+    text: str = Field(..., description="Text to be said.")
+
+class SayTool(BaseTool):
+    name: str = "say"
+    description: str = "Tool used for speaking in e.g. human-robot conversation"
+    robot: type[Robot]
+    args_schema: Type[SayToolInput] = SayToolInput
+
+    def _run(self, text: str):
+        return self.robot.speak(text)
+
+def state_retriever():
+    """State retriever used for feeding state information to agent every interation"""
+    return {"temperature": 30, "battery_state": 33}
```

Comment:
Could you include logs here? Or some other less trivial example of state?

Reply:
I've updated the docstring to provide a clearer explanation of what the state_retriever function should do. This example purposefully abstracts several implementation details and, in its current form, does not require use-case-specific features.

Although, once there is a simple log grabber implemented in rai.py, we should consider updating the example.